### ISSUE TITLE
Add a function to limit the GB filesize of the query log file

### ISSUE THREAD

**Issue by DandelionSprout** [AdguardTeam/AdGuardHome#4284]
Have a question or an idea? Please search it [on our forum](https://github.com/AdguardTeam/AdGuardHome/discussions) to make sure it was not yet asked. If you cannot find what you had in mind, please [submit it here](https://github.com/AdguardTeam/AdGuardHome/discussions/new).

### Prerequisites

Please answer the following questions for yourself before submitting an issue. **YOU MAY DELETE THE PREREQUISITES SECTION.**

- [x] I am running the latest version (v0.108.0-b.2)
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Problem Description

Currently, the only method to limit the query log size's file, is by the amount of days the queries are stored. This works for 99.9% of AdGuard Home users, but can become a potential problem for those who run [public servers with 100-ish weekly users](https://github.com/DandelionSprout/adfilt/tree/master/Dandelion%20Sprout's%20Official%20DNS%20Server). Keeping the query log time setting set to "30 days", has apparently led to an impressively large log file size of 9.2 GiB, becoming the 2nd-largest single file on my PC. 
![image](https://user-images.githubusercontent.com/22780683/153740413-20b616bd-e657-4ae5-8610-c3d76a7deb01.png)

### Proposed Solution

In addition to the amount of days queries can be stored, it should also be possible to set a max amount of GB (be it Linux-GB or GiB) or MB the file size can reach before the oldest entries are deleted.

### Alternatives Considered

None that I'm currently aware of. Though I should probably have reduced the log days on my setup from 30 to 7 days either way.

### Additional Information

The existing setting for the cache size (which I've set to 40 MB), doesn't seem to affect the query log's size.

**Comment by ainar-g** [comment]
This may be partially addressed by #3404.

 >  In addition to the amount of days queries can be stored, it should also be possible to set a max amount of GB (be it Linux-GB or GiB) or MB the file size can reach before the oldest entries are deleted.

Which of the two ways would have the priority?  I.e. if my log has reached the size limit but not the duration limit, should entries still be deleted?

**Comment by DandelionSprout** [comment]
>I.e. if my log has reached the size limit but not the duration limit, should entries still be deleted?

Correct. It'd be a "Whichever of the two that happens first" situation.

**Comment by Nahor** [comment]
I think this issue can also be seen as a security problem. If a DNS client goes rogue,  it can send a lot of requests just to fill the logs. This eventually can fill the storage device, especially on some machines like routers. Once the storage is full, things start to break down: recent queries are lost, filters don't get updated, stats are incorrect, ...

**Comment by imohitkr** [comment]
Just happened with me , the log size went up to 30 GB

**Comment by SoFMeRight** [comment]
+1 to this idea!

**Comment by siliconfire** [comment]
+1

**Comment by irvis007** [comment]
> Just happened with me , the log size went up to 30 GB

same here ✋
logsize 30+ GB,  some crap on mobile client wanted to reach malware address 7mln times in a 4 days 🌵

**Comment by Sil3ntVip3r** [comment]
The current query log remains time-rotated only.  Writes append without a byte
limit, and storage consists of the active `querylog.json` plus one rotated
`.1` file.  Multi-gigabyte logs and disk exhaustion therefore remain possible.

Before implementation, `max_size` needs one precise invariant: it should be
the aggregate query-log disk budget, not the size of each file.  A per-file
interpretation permits approximately twice the configured amount.

Suggested acceptance criteria:

- disabled by default for compatibility;
- earlier of time rotation or size enforcement;
- aggregate accounting across active and rotated files;
- bounded, documented overshoot no greater than one encoded record or flush
  batch;
- deterministic behavior when the cap is reduced at runtime;
- defined handling when one record exceeds the cap;
- safe restart and Windows rename/remove behavior;
- no corruption or race with flush, clear, or rotation.

Rewriting a multi-gigabyte JSON log on every append would not be acceptable.
The implementation should retain segmented append behavior and enforce the
budget while holding the existing file-write serialization.

Tests should use a tiny deterministic cap and cover boundary crossing,
active-plus-rotated total size, restart, time-versus-size ordering, cap
reduction, oversized records, filesystem failures, and concurrent
flush/clear/rotate under `-race`.
